### ISSUE TITLE
Stuck on saving (save) after editing token

### ISSUE THREAD

**Issue by Jean-Baptiste-Camps** [hipster-philology/pandora-postcorrect-app#18]
I seem to be stuck on "saving", when I edit a token. Here is the full log

````
(env) jbc@Hrodgang-4:~/Data/F/pandora/pandora-postcorrect-app$ python manage.prun
SECRET KEY ENV VAR NOT SET! SHOULD NOT SEE IN PRODUCTION
THIS APP IS IN DEBUG MODE. YOU SHOULD NOT SEE THIS IN PRODUCTION.
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
SECRET KEY ENV VAR NOT SET! SHOULD NOT SEE IN PRODUCTION
THIS APP IS IN DEBUG MODE. YOU SHOULD NOT SEE THIS IN PRODUCTION.
 * Debugger is active!
 * Debugger PIN: 119-894-154
127.0.0.1 - - [02/Oct/2017 11:59:47] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [02/Oct/2017 11:59:48] "GET /statics/css/bootstrap.min.css HTTP/1.1" 304 -
127.0.0.1 - - [02/Oct/2017 11:59:48] "GET /statics/css/typeaheadjs.css HTTP/1.1" 304 -
127.0.0.1 - - [02/Oct/2017 11:59:48] "GET /statics/js/popper.js HTTP/1.1" 304 -
127.0.0.1 - - [02/Oct/2017 11:59:48] "GET /statics/js/jquery.min.js HTTP/1.1" 304 -
127.0.0.1 - - [02/Oct/2017 11:59:48] "GET /statics/js/bootstrap.min.js HTTP/1.1" 304 -
127.0.0.1 - - [02/Oct/2017 11:59:48] "GET /statics/js/typeahead.js HTTP/1.1" 304 -
127.0.0.1 - - [02/Oct/2017 11:59:51] "GET /corpus/1/tokens/edit HTTP/1.1" 200 -
127.0.0.1 - - [02/Oct/2017 11:59:57] "POST /corpus/1/tokens/edit/1 HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/app/main/views/tokens.py", line 51, in tokens_edit_single
    morph=string_to_none(request.form.get("morph"))
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/app/models/linguistic.py", line 324, in update
    validity = WordToken.is_valid(lemma=lemma, POS=POS, morph=morph, corpus=corpus)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/app/models/linguistic.py", line 253, in is_valid
    and corpus.get_allowed_values("lemma", label=lemma).count() == 0:
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/app/models/linguistic.py", line 29, in get_allowed_values
    db.and_(cls.corpus == self.id, cls.label == label).order_by(cls.label)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/sqlalchemy/sql/elements.py", line 682, in __getattr__
    key)
AttributeError: Neither 'BooleanClauseList' object nor 'Comparator' object has an attribute 'order_by'

````

**Comment by PonteIneptique** [comment]
It seems I have misplaced a parenthesis. It might be super easy to fix ;)

Le 2 oct. 2017 12:01 PM, "Jean-Baptiste-Camps" <notifications@github.com> a
écrit :

**Comment by PonteIneptique** [comment]
Potential fix in @4cc5cf3

**Comment by Jean-Baptiste-Camps** [comment]
Nope, alas. Now, if I don't edit the content of lemma, I get «Invalid value» (with a right lemma). And, if I edit it, I still get stuck.

1. I think the invalid value is due to a space being added after the end of each lemma. I don't see it in my original data, so maybe it is a bug.
2. If I edit, I get another error

````
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/env/lib/python3.5/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/jbc/Data/F/pandora/pandora-postcorrect-app/app/main/views/tokens.py", line 51, in tokens_edit_single
    morph=string_to_none(request.form.get("morph"))
TypeError: 'WordToken' object is not iterable
````

**Comment by PonteIneptique** [comment]
Could you give me a small sample of your data to check with it (A sample that would not be copyright and you would not find any issue for them to be in a test :D )

**Comment by Jean-Baptiste-Camps** [comment]
Well, nothing is copyrighted, since it is only medieval texts. You can find it in Geste, but here is a small sample from Floovant,


[Floovant_echantillon.txt](https://github.com/hipster-philology/pandora-postcorrect-app/files/1351660/Floovant_echantillon.txt)


And the list of lemmas.

[Liste_lemmes.txt](https://github.com/hipster-philology/pandora-postcorrect-app/files/1351640/Liste_lemmes.txt)

**Comment by PonteIneptique** [comment]
You know what would be a great thing : add this list of lemma to the wiki for later purpose :)

**Comment by Jean-Baptiste-Camps** [comment]
Sure. Could you create the wiki so I can add it ?

**Comment by PonteIneptique** [comment]
I set you as repo owner and created the wiki ;)

**Comment by PonteIneptique** [comment]
You're spot on : there is leading space. I found it through a test writing (Gotta love tests..)

**Comment by PonteIneptique** [comment]
I need to write more tests (that include morph) but can you check https://github.com/hipster-philology/pandora-postcorrect-app/tree/issue-18 ?

**Comment by Jean-Baptiste-Camps** [comment]
Is this what you add in mind ?
https://github.com/hipster-philology/pandora-postcorrect-app/wiki/List-of-allowed-lemmas

**Comment by PonteIneptique** [comment]
Yup. That's perfect.
I'll let you eval the issue-18 branch and tell me if this is okay.
Then I'll close the issue by a PR :)

(Take a look at how I wrote the test. Might be of interest for you ;))

**Comment by Jean-Baptiste-Camps** [comment]
Same problem here (I recreated the db to be sure), 
````
TypeError: 'WordToken' object is not iterable
````

**Comment by PonteIneptique** [comment]
Damn. It could be the morph then....
And here I go writing another test and fixture..

**Comment by PonteIneptique** [comment]
Well not today actually. I started at 6.30. Let's give it a rest for the day ;)

**Comment by PonteIneptique** [comment]
I am completely unable to reproduce the bug...

**Comment by PonteIneptique** [comment]
Fixed by #24
